Add Open-Meteo solar forecast

thecem opened the pull request.

thecem committed:
adding octopus-germany tariff

added octups DE api
added user&passwd auth
added token handling
intelligent octopus go tarif -> GoPrice / StandardPrice (Manual)
arif time 0-5 Go Rate 5-24 Standard Rate
static price for nowt
o do: use price information from api (if listet)

thecem committed:
first draft of octopus-de tariff plugin
auth via email and password
pulls the first gross rate from the octopus api oeg-kraken.energy
Set the price for 24h to the gross rate, else to 23.25 cent/kWh

thecem committed:
deleted:    tariff/octopus-germany.go
deleted:    templates/definition/tariff/octopus-germany.yaml

thecem committed:
Open-Meteo Solar Forecast Tariff
Erste Version des Open-Meteo Solar Forecast Tariffs
Brauche Unterstützung wie die Daten in den Tariff kommen.

thecem committed:
Open-Meteo Solar Forcast

!WIP!  -> NEED HELP!

Problems to get the data into the tariff.

Calculations and API Working!

thecem committed:
Open-Meteo Solar Forecast

!NEED HELP!

!!WIP!!

API -> Working
WhPeriod Calculation -> Working (Includs Damping!)
Tariff Solar: not working!

andig commented:
Bitte PRs trennen!

thecem commented:
Ich habe die für octopus geschlossen und wollte diesen hier separat öffnen VSC wollte das aber nicht wie ich.
Zugehörig ist: 
https://github.com/evcc-io/evcc/pull/19005
und
https://github.com/evcc-io/evcc/pull/18940

Die sind aber nicht zu betrachten der Octopus läuft braucht aber noch anpassungen (WIP).

Hier mit dem Open-Meteo komme ich nicht weiter die Daten an die Tariff api so zu parsen das es geht. Hier brauche ich hilfe. Ich hatte zwischendrin probleme mit dem Zeitformat {parseISO8601ToRFC3339} (could not parse "" for ":") aber das ist jetzt gelöst, dafür taucht dann in der UI nicht mehr der die Vorhersage auf. Ich denke es hat was mit dem format der daten zur tariff api zu tun.

andig commented:
Unabhängig davon bitte ein PR pro Thema!

thecem commented:
also soll ich den jetzt löschen und neu?
________________________________
Von: andig ***@***.***>
Gesendet: Sonntag, 23. Februar 2025 20:08
An: evcc-io/evcc ***@***.***>
Cc: thecem ***@***.***>; Author ***@***.***>
Betreff: Re: [evcc-io/evcc] Open-Meteo Solar Forecast (PR #19111)


Unabhängig davon bitte ein PR pro Thema!

—
Reply to this email directly, view it on GitHub<https://github.com/evcc-io/evcc/pull/19111#issuecomment-2677066203>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALD42A26FLAI5OXA6KI7GVT2RIMCDAVCNFSM6AAAAABXWO5KTGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZXGA3DMMRQGM>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

[andig]andig left a comment (evcc-io/evcc#19111)<https://github.com/evcc-io/evcc/pull/19111#issuecomment-2677066203>

Unabhängig davon bitte ein PR pro Thema!

—
Reply to this email directly, view it on GitHub<https://github.com/evcc-io/evcc/pull/19111#issuecomment-2677066203>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALD42A26FLAI5OXA6KI7GVT2RIMCDAVCNFSM6AAAAABXWO5KTGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZXGA3DMMRQGM>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

andig commented:
Egal wie- getrennte prs bitte. Ein Branch und damit pr je Thema.

thecem committed:
Remove octopus-de.yaml from PR

thecem committed:
Remove octopus-de.go from PR

thecem committed:
Remove types.go from PR

thecem commented:
Ich hoffe das passt jetzt so, ich habe die jetzt rausgelöscht.

thecem commented:
Ich werde das noch stark vereinfachen. Der Vorteil von Open-Meteo ist die hohe Anzahl von Calls und die hohe Genauigkeit.
Deshalb bitte ich hier um Unterstützung wie ich die Daten in die Tarif.api bekomme.

zachelnet commented:
> Ich werde das noch stark vereinfachen. Der Vorteil von Open-Meteo ist die hohe Anzahl von Calls und die hohe Genauigkeit. Deshalb bitte ich hier um Unterstützung wie ich die Daten in die Tarif.api bekomme.

Sieht interessant aus.
Eventuell würde ich den Teil mit der Webservice-Abfrage in das Template integrieren und dann die Werte an das Model übergeben, damit die Werte entsprechend der Prognose ausgegeben werden. Dann könnte man diese Werte über das Template an
Tarif/Solcast übergeben.

Dann bräuchte man nur noch Code für die Berechnung und entsprechend das Template. Dies würde die Wartbarkeit sehr vereinfachen und eventuell könnten noch weitere APIs entsprechend angebunden werden oder alternativ das Berechnungsmodell ausgetauscht werden.

thecem commented:
Am Wochenende kommt eine neue einfach Lösung.

thecem committed:
Massiv simplification of the Open-Meteo tariff.
Only a template file is left, the rest is removed.
The Open-Meteo tariff is now a simple example for a tariff definition.

  solar:
    type: template
    template: open-meteo
    lat: 50.1592
    lon: 8.9830
    az: 0
    dec: 37
    dckwp: 125
    ackwp: 9
    eff: 1
    dm: 0
    de: 0
mandatory:
    lat: 50.1592
    lon: 8.9830
    az: 0
    dec: 37
    dckwp: 125
optional:
    ackwp: 9
    eff: 1
    dm: 0
    de: 0

thecem committed:
modified:   templates/definition/tariff/open-meteo.yaml

thecem commented:
Massiv simplification of the Open-Meteo tariff.

Only a template file is left, the rest is removed.
The Open-Meteo tariff is now a simple example for a tariff definition.

  solar:
    type: template
    template: open-meteo
    lat: 50.1592
    lon: 8.9830
    az: 0
    dec: 37
    dckwp: 125
    ackwp: 9
    eff: 1
    dm: 0
    de: 0
mandatory:
    lat: 50.1592
    lon: 8.9830
    az: 0
    dec: 37
    dckwp: 125
optional:
    ackwp: 9
    eff: 1
    dm: 0
    de: 0

thecem commented:
if you can´t wait for release and would like to test now:
 ```
tariffs:
  solar:
    type: custom
    forecast:
      source: http
      uri: https://api.open-meteo.com/v1/forecast?latitude=50.1592&longitude=8.9830&azimuth=180&tilt=37&hourly=temperature_2m,global_tilted_irradiance,global_tilted_irradiance_instant&daily=sunrise,sunset&forecast_days=3&timezone=auto&timeformat=unixtime
      jq: |
        
        def eff: 0.83; # efficiency 1 = 100%
        
        def dckwp: 135.2; # DC-kWp

        def ackwp: 9.0; # AC-kWp

        def dm: 0; # 1 = 100% damping morning (0 = no damping)

        def de: 0; # damping evening

        def clamp(min; x; max):
          if x < min then min elif x > max then max else x end;
        def midday(sunrise; sunset):
          sunrise + ((sunset - sunrise) / 2);
        def calculate_damping(time; sunrise; sunset; m; e):
          if time < sunrise then m
          elif time < midday(sunrise; sunset) then
            m * (1 - ((time - sunrise) / (midday(sunrise; sunset) - sunrise)))
          elif time < sunset then
            e * ((time - midday(sunrise; sunset)) / (sunset - midday(sunrise; sunset)))
          else
            e
          end;
        .hourly as $h
        | .daily as $d
        | [ range(0; ($h.time | length))
            | . as $i
            | $h.time[$i] as $time
            | ($i / 24 | floor) as $day
            | {
                start: ($time | todateiso8601),
                end: (($time + 3600) | todateiso8601),
                price: (
                  (dckwp * 1000)
                  * ($h.global_tilted_irradiance[$i] / 1000)
                  * (1 + (-0.00045 *
                        (
                          ( ($h.temperature_2m[$i]
                              + ($h.temperature_2m[$i-1] // $h.temperature_2m[$i])
                            ) / 2)
                          + ($h.global_tilted_irradiance[$i] / 800.0) * 0.0342 - 25.0
                        )
                      ))
                  * (1 - calculate_damping($time;
                        $d.sunrise[$day];
                        $d.sunset[$day];
                        dm; de))
                  * eff
                  | clamp(0; .; (ackwp * 1000))
                )
              }
          ]
        | tostring
```

CiNcH83 commented:
Are you sure that azimuth is [-180,180]? On HA it is [0,360]...

thecem commented:
The term azimuth refers to a horizontal angle measured relative to the cardinal directions. There are two main types of azimuth measurements:
	1.	South Azimuth: The measurement starts in the south (0°) and proceeds through west (90°), north (180°), to east (270°).
	2.	North Azimuth: This system begins in the north (0°) and moves clockwise through east (90°), south (180°), to west (270°). This method is commonly used in navigation and geodesy.
The choice of convention depends on the application, such as astronomy or navigation.

The template uses the API based south azimuth:

Total radiation received on a tilted pane as average of the preceding hour. The calculation is assuming a fixed albedo of 20% and in isotropic sky. Please specify tilt and azimuth parameter. Tilt ranges from 0° to 90° and is typically around 45°. Azimuth should be close to 0° (0° south, -90° east, 90° west). If azimuth is set to "nan", the calculation assumes a horizontal tracker. If tilt is set to "nan", it is assumed that the panel has a vertical tracker. If both are set to "nan", a bi-axial tracker is assumed.

https://open-meteo.com/en/docs

GrimmiMeloni commented:
Fixed the linter here locally (a new line is missing at the end of the template), but a PR leads to a huge changeset. I am not that good at working with git "triangles". :(

thecem committed:
Fixed linter last new line

GrimmiMeloni commented:
You may want to run `make lint-ui` locally, to recreate the checks that the GHA does.

thecem committed:
fix lint with make lint-ui

thecem committed:
Merge branch 'evcc-io:master' into master

tantive commented:
@thecem Is there a reason why 'global_tilted_irradiance_instance' is fetched? The calculation does not seem to use it.

thecem commented:
You are rigt this is at the moment not nessesary.

zachelnet commented:
Hi @thecem thanks for the code, looks good, without the optional value works.

During my test with the following configuration with east and west orientation, regarding the option `dm` and `de`, the following error messages appear.

config:
```
- type: template
  # Ost 
  template: open-meteo
  lat: 48.768210
  lon: 10.799644
  dec: 45
  az: -74
  dckwp: 4.3
  de: 0.5
- type: template
  # West
  #template: forecast-solar
  template: open-meteo
  lat: 48.768210
  lon: 10.799644
  dec: 45
  az: 74
  dckwp: 4.3
  dm: 0.5
```

log messages:
```
[main ] ERROR 2025/03/02 18:10:11 creating tariff solar-template-1 failed: cannot create tariff type 'template': cannot create tariff type 'custom': forecast: invalid jq query 'def eff: 1; # efficiency 1 = 100% def dckwp: 4.3 * 1000 ; # DC-kWp def ackwp: 1000 * 1000 ; # AC-kWp def dm: '0.5'; # 1 = 100% damping morning (0 = no damping) def de: 0; # damping evening def clamp(min; x; max): if x < min then min elif x > max then max else x end; def midday(sunrise; sunset): sunrise + ((sunset - sunrise) / 2); def calculate_damping(time; sunrise; sunset; m; e): if time < sunrise then m elif time < midday(sunrise; sunset) then m * (1 - ((time - sunrise) / (midday(sunrise; sunset) - sunrise))) elif time < sunset then e * ((time - midday(sunrise; sunset)) / (sunset - midday(sunrise; sunset))) else e end; .hourly as $h | .daily as $d | [ range(0; ($h.time | length)) | . as $i | $h.time[$i] as $time | ($i / 24 | floor) as $day | { start: ($time | todateiso8601), end: (($time + 3600) | todateiso8601), price: ( dckwp * ($h.global_tilted_irradiance[$i] / 1000) * (1 + (-0.00045 * ( ( ($h.temperature_2m[$i] + ($h.temperature_2m[$i-1] // $h.temperature_2m[$i]) ) / 2) + ($h.global_tilted_irradiance[$i] / 800.0) * 0.0342 - 25.0 ) )) * (1 - calculate_damping($time; $d.sunrise[$day]; $d.sunset[$day]; dm; de)) * eff | clamp(0; .; ackwp) ) } ] | tostring ': unexpected token "'"
[main ] ERROR 2025/03/02 18:10:11 creating tariff solar-template-0 failed: cannot create tariff type 'template': cannot create tariff type 'custom': forecast: invalid jq query 'def eff: 1; # efficiency 1 = 100% def dckwp: 4.3 * 1000 ; # DC-kWp def ackwp: 1000 * 1000 ; # AC-kWp def dm: 0; # 1 = 100% damping morning (0 = no damping) def de: '0.5'; # damping evening def clamp(min; x; max): if x < min then min elif x > max then max else x end; def midday(sunrise; sunset): sunrise + ((sunset - sunrise) / 2); def calculate_damping(time; sunrise; sunset; m; e): if time < sunrise then m elif time < midday(sunrise; sunset) then m * (1 - ((time - sunrise) / (midday(sunrise; sunset) - sunrise))) elif time < sunset then e * ((time - midday(sunrise; sunset)) / (sunset - midday(sunrise; sunset))) else e end; .hourly as $h | .daily as $d | [ range(0; ($h.time | length)) | . as $i | $h.time[$i] as $time | ($i / 24 | floor) as $day | { start: ($time | todateiso8601), end: (($time + 3600) | todateiso8601), price: ( dckwp * ($h.global_tilted_irradiance[$i] / 1000) * (1 + (-0.00045 * ( ( ($h.temperature_2m[$i] + ($h.temperature_2m[$i-1] // $h.temperature_2m[$i]) ) / 2) + ($h.global_tilted_irradiance[$i] / 800.0) * 0.0342 - 25.0 ) )) * (1 - calculate_damping($time; $d.sunrise[$day]; $d.sunset[$day]; dm; de)) * eff | clamp(0; .; ackwp) ) } ] | tostring ': unexpected token "'"
```

thecem committed:
remove: global_tilted_irradiance_instance

thecem commented:
Many thanks for heads-up! Could you do me a favor and try it without the zero in front of the dot:

```
- type: template
  # East 
  template: open-meteo
  lat: 48.768210
  lon: 10.799644
  dec: 45
  az: -74
  dckwp: 4.3
  de: .5
```

and with a 1 in front of?
```
- type: template
  # East 
  template: open-meteo
  lat: 48.768210
  lon: 10.799644
  dec: 45
  az: -74
  dckwp: 4.3
  de: 1.5
```
What is the result?

zachelnet commented:
Hi @thecem 
> ```
> - type: template
>   # East 
>   template: open-meteo
>   lat: 48.768210
>   lon: 10.799644
>   dec: 45
>   az: -74
>   dckwp: 4.3
>   de: .5
> ```

Doesn't work with `.5':
```
[main ] ERROR 2025/03/02 19:47:37 creating tariff solar-template-0 failed: cannot create tariff type 'template': cannot create tariff type 'custom': forecast: invalid jq query 'def eff: 1; # efficiency 1 = 100% def dckwp: 4.3 * 1000 ; # DC-kWp def ackwp: 8 * 1000 ; # AC-kWp def dm: 0; # 1 = 100% damping morning (0 = no damping) def de: '0.5'; # damping evening def clamp(min; x; max): if x < min then min elif x > max then max else x end; def midday(sunrise; sunset): sunrise + ((sunset - sunrise) / 2); def calculate_damping(time; sunrise; sunset; m; e): if time < sunrise then m elif time < midday(sunrise; sunset) then m * (1 - ((time - sunrise) / (midday(sunrise; sunset) - sunrise))) elif time < sunset then e * ((time - midday(sunrise; sunset)) / (sunset - midday(sunrise; sunset))) else e end; .hourly as $h | .daily as $d | [ range(0; ($h.time | length)) | . as $i | $h.time[$i] as $time | ($i / 24 | floor) as $day | { start: ($time | todateiso8601), end: (($time + 3600) | todateiso8601), price: ( dckwp * ($h.global_tilted_irradiance[$i] / 1000) * (1 + (-0.00045 * ( ( ($h.temperature_2m[$i] + ($h.temperature_2m[$i-1] // $h.temperature_2m[$i]) ) / 2) + ($h.global_tilted_irradiance[$i] / 800.0) * 0.0342 - 25.0 ) )) * (1 - calculate_damping($time; $d.sunrise[$day]; $d.sunset[$day]; dm; de)) * eff | clamp(0; .; ackwp) ) } ] | tostring ': unexpected token "'"
```

> and with a 1 in front of?
> 
> ```
> - type: template
>   # East 
>   template: open-meteo
>   lat: 48.768210
>   lon: 10.799644
>   dec: 45
>   az: -74
>   dckwp: 4.3
>   de: 1.5
> ```
> 
> What is the result?

The result with '1.5' it's work:
```
[solar-template-0] TRACE 2025/03/02 19:49:21 GET https://api.open-meteo.com/v1/forecast?latitude=48.76821&longitude=10.799644&azimuth=-74&tilt=45&hourly=temperature_2m,global_tilted_irradiance&daily=sunrise,sunset&forecast_days=3&timezone=auto&timeformat=unixtime
```

thecem committed:
Changed eff, dm and de to 0 - 100 insted of a 0 to 1 value.
The received values are in some cases stings instead of numeric values.
This is a temporary fix until the issue is resolved.
When a 0.5 value is received, the value is in some cases string.
I think this belongs to a problem in the class_enummer.go file.
The values are not converted to float64 in the class_enummer.go file.

thecem commented:
There is no problem with the script, but a problem with the type of input values in the template mechanism.

I fixed it with a workaround. Now you have to set 0-100 values.

zachelnet commented:
@thecem thanks looks good

tantive commented:
@thecem Does it make sense to allow configuration of other constants used in the calculcation e.g. temperature coefficient of the used modules?

CiNcH83 commented:
@zachelnet I assume you have one 8 kW inverter? You specified an 8 kW inverter for each orientation. Guess it won't harm too much in this case. But probably it is better to omit AC power in such a case as it is optional anyway?

thecem committed:
Add OctopusEnergy DE tariff.
Integrating DE API OctopusEnergy tariff for Germany.
SimpleProduct and TimeOfUseProduct are supported.
In case of TimeOfUseProduct, the tariff shows the rates for the TimeOfUseProduct.
In case of SimpleProduct, the tariff shows the rates for the SimpleProduct for 72 hours.
If both SimpleProduct and TimeOfUseProduct are not available, the tariff shows the rates of 0.2827.
The intelligent dispatching of the tariff will be added in the next PR.
The tariff is tested with the OctopusEnergy DE API.

thecem committed:
removing octopus-de tariff should be an exta PR.

thecem committed:
added function for rossmodel and alphatemp

thecem committed:
Addapted documentation in open-meteo.yaml

andig commented on templates/definition/tariff/open-meteo.yaml:
```diff
@@ -46,39 +46,39 @@ params:
-      en: AC-kWp
-      de: AC-kWp
+      en: AC-kWp Inverter power
+      de: AC-kWp Wechselrichterleistung
```
```suggestion
      de: AC Leistung (kW)
```

andig commented on templates/definition/tariff/open-meteo.yaml:
```diff
@@ -0,0 +1,163 @@
+  - name: dckwp
+    description:
+      en: DC-kWp
+      de: DC-kWp
```
```suggestion
      de: DC Maximalleistung (kWp)
```

thecem committed:
Update templates/definition/tariff/open-meteo.yaml

ja besser!

Co-authored-by: andig <cpuidle@gmail.com>

thecem committed:
Update templates/definition/tariff/open-meteo.yaml

auch!

Co-authored-by: andig <cpuidle@gmail.com>

thecem commented on templates/definition/tariff/open-meteo.yaml:
```diff
@@ -0,0 +1,163 @@
+      de: Dämpfung abends (0 = keine Dämpfung 100 = volle Dämpfung)
+    default: 0
+    advanced: true
+  - name: eff
```
absolut

thecem committed:
Update templates/definition/tariff/open-meteo.yaml

Co-authored-by: andig <cpuidle@gmail.com>

thecem commented on templates/definition/tariff/open-meteo.yaml:
```diff
@@ -0,0 +1,163 @@
+    example: -0.004
+    default: -0.004
+    advanced: true
+  - name: rossmodel
```
don´t get you? what exactly?

andig replied:
all the roos models are just float values. Where do they come from? Can we have the user just enter the float value he needs? Its advanced anyway?

thecem replied:
https://www.sciencedirect.com/science/article/pii/S2772671122000754

thecem committed:
Requested review changes to the open-meteo.yaml file

andig commented on templates/definition/tariff/open-meteo.yaml:
```diff
@@ -0,0 +1,163 @@
+    advanced: true
+  - name: rossmodel
+    description:
+      en: Cooling type of pv module [Ross Model]
```
there's another one

thecem replied:
deleted!

thecem committed:
change rossmodel to float
deleted repeating pv module
added paper link for the rossmodel
added help text for the rossmodel

andig commented:
Great PR, thank you. Just be aware that you'll need to maintain this until the end- my jq-foo is not up for this ;)

andig commented:
@thecem please see https://github.com/evcc-io/evcc/discussions/19463#discussioncomment-12411914
